flink-streaming-java/src/test/java/org/apache/flink/streaming/api/DataStreamTest.java

```diff
@@ -906,6 +919,256 @@ public void testChannelSelectors() {
+				new KeySelector<Tuple2<Integer[], String>, int[]>() {
+
+			@Override
+			public int[] getKey(Tuple2<Integer[], String> value) throws Exception {
+				int[] ks = new int[value.f0.length];
+				for (int i = 0; i < ks.length; i++) {
+					ks[i] = value.f0[i];
+				}
+				return ks;
+			}
+		};
+
+		testKeyRejection(keySelector, PrimitiveArrayTypeInfo.INT_PRIMITIVE_ARRAY_TYPE_INFO);
+	}
+
+	@Test
+	public void testBasicArrayKeyRejection() {
+
+		KeySelector<Tuple2<Integer[], String>, Integer[]> keySelector =
+				new KeySelector<Tuple2<Integer[], String>, Integer[]>() {
+
+			@Override
+			public Integer[] getKey(Tuple2<Integer[], String> value) throws Exception {
+				return value.f0;
+			}
+		};
+
+		testKeyRejection(keySelector, BasicArrayTypeInfo.INT_ARRAY_TYPE_INFO);
+	}
+
+	@Test
+	public void testObjectArrayKeyRejection() {
+
+		KeySelector<Tuple2<Integer[], String>, TestClass[]> keySelector =
+				new KeySelector<Tuple2<Integer[], String>, TestClass[]>() {
+
+					@Override
+					public TestClass[] getKey(Tuple2<Integer[], String> value) throws Exception {
+						TestClass[] ks = new TestClass[value.f0.length];
+						for (int i = 0; i < ks.length; i++) {
+							ks[i] = new TestClass(value.f0[i]);
+						}
+						return ks;
+					}
+				};
+
+		ObjectArrayTypeInfo<TestClass[], TestClass> keyTypeInfo = ObjectArrayTypeInfo.getInfoFor(
+				TestClass[].class, new GenericTypeInfo<>(TestClass.class));
+
+		testKeyRejection(keySelector, keyTypeInfo);
+	}
+
+	private <K> void testKeyRejection(KeySelector<Tuple2<Integer[], String>, K> keySelector, TypeInformation<K> expectedKeyType) {
+
+		StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
+
+		DataStream<Tuple2<Integer[], String>> input = env.fromElements(
+				new Tuple2<>(new Integer[] {1, 2}, "barfoo")
+		);
+
+		Assert.assertEquals(expectedKeyType, TypeExtractor.getKeySelectorTypes(keySelector, input.getType()));
+
+		// adjust the rule
+		expectedException.expect(InvalidProgramException.class);
+		expectedException.expectMessage("This type (" + expectedKeyType + ") cannot be used as key.");
+
+		input.keyBy(keySelector);
+	}
+
+	// composite key tests : POJOs
+
+	@Test
+	public void testPOJONestedArrayKeyRejection() {
+		StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
+
+		DataStream<POJOwithHashCode> input = env.fromElements(
+				new POJOwithHashCode(new int[] {1, 2}));
+
+		TypeInformation<?> expectedTypeInfo = new TupleTypeInfo<Tuple1<int[]>>(
+				PrimitiveArrayTypeInfo.INT_PRIMITIVE_ARRAY_TYPE_INFO);
+
+		// adjust the rule
+		expectedException.expect(InvalidProgramException.class);
+		expectedException.expectMessage("This type (" + expectedTypeInfo + ") cannot be used as key.");
+
+		input.keyBy("id");
+	}
+
+	@Test
+	public void testNestedArrayWorkArround() {
+		StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
+
+		DataStream<POJOwithHashCode> input = env.fromElements(
```

Comment:
The focus of this test appears to be to verify that you can use an array as a key by wrapping it in a Pojo that implements HashCode. (based on the naming).

We should probably focus more on the Pojo-with-hashCode implementation part; doesn't really matter what is contained. We are never checking that anyway.